### ISSUE TITLE
vectorization regression in Clang 19

### ISSUE THREAD

**Issue by eyelash** [llvm/llvm-project#114856]
When compiling the following code on x86-64 with `-O2`

```cpp
void foo(float* result, int size, float y, float delta) {
    for (int t = 0; t < size; ++t) {
        result[t] = y + delta * t;
    }
}
```

Clang generates a lot of code that seems suboptimal. After [asking on Stack Overflow](https://stackoverflow.com/q/79156564/799849) I was told that this is a regression in Clang 19 that I should report here. The comments and [this answer](https://stackoverflow.com/a/79156703/799849) on Stack Overflow shed more light into what's going on.

[Here's a link to this example on Compiler Explorer with Clang trunk](https://godbolt.org/z/7Y67cKTfx) that shows the suboptimal assembly.

**Comment by goldsteinn** [comment]
> @goldsteinn This looks like fallout from sitofp to uitofp canonicalization. Presumably we don't have undo code for x86 yet?

We do: https://github.com/llvm/llvm-project/commit/6c40d463c28e7a6843bea9f6d838cd89e586cbe8

I think the issue is we are dropping the `nneg` flag from `uitofp` when vectorizing in middle-end: https://godbolt.org/z/q97xvvsM6

**Comment by goldsteinn** [comment]
> > @goldsteinn This looks like fallout from sitofp to uitofp canonicalization. Presumably we don't have undo code for x86 yet?
> 
> We do: [6c40d46](https://github.com/llvm/llvm-project/commit/6c40d463c28e7a6843bea9f6d838cd89e586cbe8)
> 
> I think the issue is we are dropping the `nneg` flag from `uitofp` when vectorizing in middle-end: https://godbolt.org/z/q97xvvsM6

Issue here seems to be with loop unroll.

**Comment by goldsteinn** [comment]
Seems like generally VPlan doesn't propagate ir flags.

**Comment by goldsteinn** [comment]
I think the fix is something along the lines of:
```
@@ -1484,6 +1484,8 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
   Value *Cast = Builder.CreateCast(Instruction::CastOps(Opcode), A, DestTy);
   State.set(this, Cast);
   State.addMetadata(Cast, cast_or_null<Instruction>(getUnderlyingValue()));
+  if (auto *CastOp = dyn_cast<Instruction>(Cast))
+    setFlags(CastOp);
 }
 
 InstructionCost VPWidenCastRecipe::computeCost(ElementCount VF,
```

Testing it, will have PR up soon if all goes well.
